### ISSUE TITLE
EXIF Data and Datestamp Overlay

### ISSUE THREAD

**Issue by prpghandi** [cordova-plugin-camera-preview/cordova-plugin-camera-preview#349]
I would like to have two features added to the plugin and would like to offer payment or donation for this work.

1. An option for the EXIF data to be made available in the file.
2. An option to overlay the date/time from the EXIF data on the photo.

Please let me know if this is possible and we can discuss the details.

**Comment by sikucing** [comment]
maybe you can use piexifjs library for that

**Comment by prpghandi** [comment]
I agree. We did use a piexifjs and it works fine. The only issue is we can't get white balance, aperture, exposure, focus info, etc... 

Do you know of anyway to get this extra info from the hardware?

I appreciate the feedback.

**Comment by davidquon** [comment]
> An option for the EXIF data to be made available in the file.

I'm interested in this as well.  @westonganger would this be something you would accept in `cordova-plugin-camera-preview`?

> An option to overlay the date/time from the EXIF data on the photo.

While I'm interested in this as well it seems outside the scope of `cordova-plugin-camera-preview` and might be better suited for a separate image viewing entity?  Just my thoughts as I'm not opposed to it but I'm not sure how it would fit into `cordova-plugin-camera-preview`

**Comment by westonganger** [comment]
Yes thats outside of the scope of this plugin. 

If the processing is done in JS and it reads base64 data then it probably should be written as a separate plugin which could be used within the success function of `takePicture`. We could recommend a solution using that in the README if this plugin ever comes to fruition.

**Comment by davidquon** [comment]
> An option for the EXIF data to be made available in the file.

Would this be something you would accept in this plugin @westonganger?  I assume so but just want to make sure you don't have any differing opinions.

**Comment by westonganger** [comment]
Well I was thinking this needed to be done on the native side of things to added within this plugin. If this is done in JS it would likely cause unnecesary performance implications and dependencies for people.

**Comment by davidquon** [comment]
I was able to get a lot of the EXIF camera data from setting `disableExifHeaderStripping` to `true`.  The EXIF data from the camera gets lost when the image is automatically rotated when `disableExifHeaderStripping` is `false`.

In the future maybe the image shouldn't be automatically rotated by default as the EXIF data should contain enough information to rotate the image properly in the image viewer or in image post processing.

**Camera EXIF:**

Key | Value
---- | -----
Make | samsung
Model | SM-G930V
Exposure | 1/180
Aperture | 1.7
Focal Length | 4.2 mm
ISO Speed | 50
Flash | No Flash

**Additional EXIF:**

Key | Value
---- | -----
ImageWidth | 1920
ImageHeight | 1080
Make | samsung
Model | SM-G930V
Orientation | Rotate 90 CW
XResolution | 72
YResolution | 72
ResolutionUnit | inches
Software | G930VVRS4BRA1
ModifyDate | 2018:03:22 11:50:44
YCbCrPositioning | Centered
ExposureTime | 1/180
FNumber | 1.7
ExposureProgram | Program AE
ISO | 50
ExifVersion | 0220
DateTimeOriginal | 2018:03:22 11:50:44
CreateDate | 2018:03:22 11:50:44
ComponentsConfiguration | Y, Cb, Cr, -
ShutterSpeedValue | 1/180
ApertureValue | 1.7
BrightnessValue | 3
ExposureCompensation | 0
MaxApertureValue | 1.7
MeteringMode | Average
LightSource | Unknown
Flash | No Flash
FocalLength | 4.2 mm
UserComment | ASCII
FlashpixVersion | 0100
ColorSpace | sRGB
ExifImageWidth | 1920
ExifImageHeight | 1080
SensingMethod | One-chip color area
SceneType | Directly photographed
ExposureMode | Auto
WhiteBalance | Auto
FocalLengthIn35mmFormat | 26 mm
SceneCaptureType | Standard
ImageUniqueID | C12QSJK01S
InteropIndex | R98 - DCF basic file (sRGB)
GPSVersionID | 2.2.0.0
Compression | JPEG (old-style)
ThumbnailOffset | 6200
ThumbnailLength | 10928

**Comment by davidquon** [comment]
Also I used `piexifjs` for adding new EXIF fields in post processing like these.

Key | Value
---- | -----
Software | AppName (Version: 0.0.1)
Artist | userName
CameraSerialNumber | deviceID
GPSLatitudeRef | North
GPSLatitude | 32.968406
GPSLongitudeRef | West
GPSLongitude | 117.035997
GPSAltitudeRef | Above Sea Level
GPSAltitude | 117 m
GPSDateStamp | 2018:03:22 11:50:45

**Comment by westonganger** [comment]
If you think that the default should be `disableExifHeaderStripping` set to true by default then I'm open to a PR for this.

If you do make a PR, please change this option name to `stripExifData` and it should default to `false`

**Comment by davidquon** [comment]
Yeah that makes more sense to me as well.  It's a little confusing as striping the exif data is combined with the rotating of the image.  I'll give it some thought but I think your suggestion might be the best compromise.  👍

**Comment by westonganger** [comment]
You could possibly separate the rotation to another option that defaults to on.

**Comment by davidquon** [comment]
> You could possibly separate the rotation to another option that defaults to on.

Yeah this is ideally how it should work.  Not sure if the default should be on though as it would increase the processing time of acquiring the image.  The problem here is the rotation causes the loss of the EXIF data currently and a mechanism would need to be added to save the EXIF data, rotate the image, and re-add the EXIF data all of which adds to the time of acquiring the image.

I'm doing something similar on the JS side and it's causing about 2-3 second delay in processing time.  IMO the defaults of the plugin should be to acquire the image expediently and if anything wants to be added slowing down acquiring the image it should be an option that can be enabled.

**Comment by keli13** [comment]
@davidquon any chance you tried running the same play on IOS? I seem to get back some reasonable data from an android device but on IOS it seems like the exif data is stripped based on what I am reading [here](https://stackoverflow.com/questions/16297730/image-upload-from-iphone-strips-exif-data?utm_medium=organic&utm_source=google_rich_qa&utm_campaign=google_rich_qa). 

I am not sure this exactly applies to this cordova plugin since the calls are delegated to the their native calls.

Thanks

**Comment by davidquon** [comment]
@keli13 - Unfortunately right now I'm only working on the Android side.  To make it work I needed to set `disableExifHeaderStripping: true` as this disables the automatic rotation of the image which as a side effect removes the exif data.  Be mindful that anything manipulating the image or converting the image will lose the exif data if not saved beforehand and written back afterwards.

I agree it sounds like Apple maybe doing something to strip the exif data.  I would just check for the exif data when getting the data back from the camera, check again before sending the file for upload, and obviously again after it's been uploaded to see if you can narrow down where it is getting lost on the iOS side.

**Comment by mm108** [comment]
@prpghandi , would you be able to share how you used piexifjs to get the exif data? I have been trying for a while and running into different sorts of problems extracting the exif data from the takePicture success function. I even tried with `window.resolveLocalFileSystemURL` and adding the `'file://'` prefix to the filePath etch. I tried logging the error and this is what I get` Error: 'load' gots invalid file data`. I am further processing the file which works fine, so there is no issue with the image file. I have set `storeToFile: true`. I just need to read the resolution / DPI.

Thanks.

> I agree. We did use a piexifjs and it works fine. The only issue is we can't get white balance, aperture, exposure, focus info, etc...
> 
> Do you know of anyway to get this extra info from the hardware?
> 
> I appreciate the feedback.

**Comment by mm108** [comment]
> I was able to get a lot of the EXIF camera data from setting `disableExifHeaderStripping` to `true`. The EXIF data from the camera gets lost when the image is automatically rotated when `disableExifHeaderStripping` is `false`.
> 
> In the future maybe the image shouldn't be automatically rotated by default as the EXIF data should contain enough information to rotate the image properly in the image viewer or in image post processing.
> 
> **Camera EXIF:**
> 
> Key	Value
> Make	samsung
> Model	SM-G930V
> Exposure	1/180
> Aperture	1.7
> Focal Length	4.2 mm
> ISO Speed	50
> Flash	No Flash
> **Additional EXIF:**
> 
> Key	Value
> ImageWidth	1920
> ImageHeight	1080
> Make	samsung
> Model	SM-G930V
> Orientation	Rotate 90 CW
> XResolution	72
> YResolution	72
> ResolutionUnit	inches
> Software	G930VVRS4BRA1
> ModifyDate	2018:03:22 11:50:44
> YCbCrPositioning	Centered
> ExposureTime	1/180
> FNumber	1.7
> ExposureProgram	Program AE
> ISO	50
> ExifVersion	0220
> DateTimeOriginal	2018:03:22 11:50:44
> CreateDate	2018:03:22 11:50:44
> ComponentsConfiguration	Y, Cb, Cr, -
> ShutterSpeedValue	1/180
> ApertureValue	1.7
> BrightnessValue	3
> ExposureCompensation	0
> MaxApertureValue	1.7
> MeteringMode	Average
> LightSource	Unknown
> Flash	No Flash
> FocalLength	4.2 mm
> UserComment	ASCII
> FlashpixVersion	0100
> ColorSpace	sRGB
> ExifImageWidth	1920
> ExifImageHeight	1080
> SensingMethod	One-chip color area
> SceneType	Directly photographed
> ExposureMode	Auto
> WhiteBalance	Auto
> FocalLengthIn35mmFormat	26 mm
> SceneCaptureType	Standard
> ImageUniqueID	C12QSJK01S
> InteropIndex	R98 - DCF basic file (sRGB)
> GPSVersionID	2.2.0.0
> Compression	JPEG (old-style)
> ThumbnailOffset	6200
> ThumbnailLength	10928

@davidquon , You mentioned that you were able to extract the exif data. Can I know what was your method to get the exif data? Thank you.
